### DISCUSSION TITLE
Is sharding already possible?

### DISCUSSION THREAD

**Discussion by cambiph** [leancodepl/patrol#1789] [Q&A] [answered]
Hey,

I read through the docs, PR's and discussions but it's a bit unclear to me what the current state is.
Is there already a way to use sharding on Firebase?
We are hitting the 60m limit so I am looking for a solution :)

Thanks in advance.
Cheers

**Comment by bartekpacia** [comment]
Sharding is possible and should work out of the box. We tested it and it works on Firebase, emulator.wtf, BrowserStack.

Check out [our example app](https://github.com/leancodepl/patrol/tree/master/packages/patrol/example) - you can try running it on FTL as a demo.

**Reply by cambiph** [reply]
So to have sharding enabled, I do not have to do anything? Because now I am running the trests as following, and sharding does not seem to be done:
```shell
gcloud firebase test android run 
--device model=MediumPhone.arm,version=33 
--type instrumentation 
--use-orchestrator 
--environment-variables clearPackageData=true 
--app build/app/outputs/apk/debug/app-debug.apk 
--test build/app/outputs/apk/androidTest/debug/app-debug-androidTest.apk 
--results-bucket #{results_bucket} 
--results-dir #{results_dir} 
--format=json 1>#{output_file} 
--timeout 60m
```

**Reply by bartekpacia** [reply]
By default `gcloud firebase test android` doesn't use sharding. Consult docs of this command here: https://cloud.google.com/sdk/gcloud/reference/beta/firebase/test/android/run#--num-uniform-shards

**Reply by cambiph** [reply]
I saw that but we need the beta SDK for that. Guess there's no other way at the moment?

**Reply by bartekpacia** [reply, marked as answer]
Maybe you could try `curl`ing Google Cloud's REST endpoints. But I think using beta is the simplest solution.

**Reply by cambiph** [reply]
I checked Flank and I am experimenting with Fladle (https://runningcode.github.io/fladle). The only issue left is that the way `MainActivityTest` is setup, fladle sees it as one test and complains about it:
`WARNING: num-uniform-shards (5) is higher than number of test cases (1) from gs://test-lab-h3pxbib2yh4f0-y4jh297cm3nvx/2023-10-12_13-01-35.953242_QrKj/app-debug-androidTest.apk`

I think if we can find a way to split the tests in `MainActivityTest`, it will run fine.
Can this be easily achieved, you think?

**Reply by bartekpacia** [reply]
I think it might be a problem on Flank's/Fladle's side. Our MainActivityTest is a valid Android instrumented test.

> I think if we can find a way to split the tests in MainActivityTest, it will run fine.  Can this be easily achieved, you think?

Achieved - maybe, easily - not so sure. MainActivityTest is set up the way it is to allow for fixing one of the most annoying problems with the integration_test plugin – https://github.com/flutter/flutter/issues/115751.

**Reply by cambiph** [reply]
Thanks for the feedback. I think I will just split up our tests in logical groups and run them in parallel on our CI. This will have the least impact and the result should be the same.

Thank you for Patrol and your amazing efforts!

**Reply by cambiph** [reply]
Mmmm, I cannot seem to find a way to split up the tests to run in parallel on Firebase as the `gcloud` command does not support the `-t integrations_tests/...` flag. Is there a workaround for this?

**Reply by bartekpacia** [reply]
I think you'd have to split tests at build time – that is, when you call `patrol build`. Build a few APKs and then upload them with separate gcloud cli invocations.

Why not just use gcloud beta, btw?

**Comment by bs-ashwanipathak** [comment]
Hey, please check out our documentation on test sharding with BrowserStack App Automate, for ref.: https://www.browserstack.com/docs/app-automate/espresso/test-sharding

**Reply by bartekpacia** [reply]
Hi, how is that information relevant to this Discussion?

**Comment by cambiph** [comment]
The cloud beta is not installed on our build agents and I cannot seem to
find the docs telling me how to install the beta version. I also cannot
find it in the gcloud-sdk-installer bucket they provide.

Op vr 13 okt. 2023 11:53 schreef Bartek Pacia ***@***.***>:

**Reply by bartekpacia** [reply]
I think you can just do `gcloud beta`:

```
$ gcloud beta
You do not currently have this command group installed.  Using it
requires the installation of components: [beta]


Your current Google Cloud CLI version is: 439.0.0
Installing components from version: 439.0.0

┌─────────────────────────────────────────────┐
│     These components will be installed.     │
├──────────────────────┬────────────┬─────────┤
│         Name         │  Version   │   Size  │
├──────────────────────┼────────────┼─────────┤
│ gcloud Beta Commands │ 2023.07.14 │ < 1 MiB │
└──────────────────────┴────────────┴─────────┘

For the latest full release notes, please visit:
  https://cloud.google.com/sdk/release_notes

Do you want to continue (Y/n)?  y

... omitted
```

**Reply by bartekpacia** [reply]
or maybe run `gcloud beta firebase test android run`, it should "just work"

**Reply by cambiph** [reply]
Already tried just running it and then it complains that it does not recognize the beta flags. Let me try if I can install it in some sort of way.

**Reply by cambiph** [reply]
Switched to Browserstack, works like a charm :)

**Reply by bartekpacia** [reply]
Happy to hear that!

Were [these docs](https://patrol.leancode.co/integrations/browserstack) useful to you in this process?

**Comment by cambiph** [comment]
Yes, I used them.
Only thing to note is that you need to manually install jq before you can
use ns_android. Other than that, it works like a charm, kiddos!

Op vr 13 okt. 2023 18:32 schreef Bartek Pacia ***@***.***>:
